datafusion/spark/src/function/string/format_string.rs

```diff
@@ -0,0 +1,2374 @@
+            'D' => Ok(TimeFormat::DUpper),
+            'F' => Ok(TimeFormat::FUpper),
+            'c' => Ok(TimeFormat::CLower),
+            _ => exec_err!("Invalid time format: {}", value),
+        }
+    }
+}
+
+impl ConversionType {
+    pub fn is_string(&self) -> bool {
+        matches!(
+            self,
+            ConversionType::StringLower | ConversionType::StringUpper
+        )
+    }
+
+    pub fn validate(&self, arg_type: DataType) -> Result<()> {
+        match self {
+            ConversionType::BooleanLower | ConversionType::BooleanUpper => {
+                if !matches!(arg_type, DataType::Boolean) {
+                    return exec_err!(
+                        "Invalid argument type for boolean conversion: {:?}",
+                        arg_type
+                    );
+                }
+            }
+            ConversionType::CharLower | ConversionType::CharUpper => {
+                if !matches!(
+                    arg_type,
+                    DataType::Int8
+                        | DataType::UInt8
+                        | DataType::Int16
+                        | DataType::UInt16
+                        | DataType::Int32
+                        | DataType::UInt32
+                        | DataType::Int64
+                        | DataType::UInt64
+                ) {
+                    return exec_err!(
+                        "Invalid argument type for char conversion: {:?}",
+                        arg_type
+                    );
+                }
+            }
+            ConversionType::DecInt
+            | ConversionType::OctInt
+            | ConversionType::HexIntLower
+            | ConversionType::HexIntUpper => {
+                if !arg_type.is_integer() {
+                    return exec_err!(
+                        "Invalid argument type for integer conversion: {:?}",
+                        arg_type
+                    );
+                }
+            }
+            ConversionType::SciFloatLower
+            | ConversionType::SciFloatUpper
+            | ConversionType::DecFloatLower
+            | ConversionType::CompactFloatLower
+            | ConversionType::CompactFloatUpper
+            | ConversionType::HexFloatLower
+            | ConversionType::HexFloatUpper => {
+                if !arg_type.is_numeric() {
+                    return exec_err!(
+                        "Invalid argument type for float conversion: {:?}",
+                        arg_type
+                    );
+                }
+            }
+            ConversionType::TimeLower(_) | ConversionType::TimeUpper(_) => {
+                if !arg_type.is_temporal() {
+                    return exec_err!(
+                        "Invalid argument type for time conversion: {:?}",
+                        arg_type
+                    );
+                }
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+}
+
+fn take_conversion_specifier(
+    s: &str,
```

Comment:
```suggestion
    mut s: &str,
```

We can specify arguments as `mut` to avoid needing to rebind below as mut:

```rust
let mut s = s;
```